### ISSUE TITLE
Unable to build keepalived-1.3.5-16_el7 from source, on RHEL 7.3 Maipo

### ISSUE THREAD

**Issue by springleo** [acassen/keepalived#1520]
I am trying to build keepalived-1.3.5-16_el7 from source, on RHEL 7.3 Maipo.
Ensured that all the underlying needed packages are installed on the build server.
cloned git repo from  acassen / keepalived .

Following are the steps i am following:

git checkout tags/v1.3.5

./configure

Keepalived configuration
------------------------
Keepalived version       : 1.3.5
Compiler                 : gcc
Preprocessor flags       :  -I/usr/include/libnl3
Compiler flags           : -Wall -Wunused -Wstrict-prototypes -Wextra -g -O2
Linker flags             :
Extra Lib                : -lcrypto -lssl -lnl-genl-3 -lnl-3 -lip4tc -lip6tc -lxtables -ldl
Use IPVS Framework       : Yes
IPVS use libnl           : Yes
IPVS syncd attributes    : No
IPVS 64 bit stats        : No
fwmark socket support    : Yes
Use VRRP Framework       : Yes
Use VRRP VMAC            : Yes
Use VRRP authentication  : Yes
With ip rules/routes     : Yes
SNMP vrrp support        : No
SNMP checker support     : No
SNMP RFCv2 support       : No
SNMP RFCv3 support       : No
DBUS support             : No
SHA1 support             : No
Use Debug flags          : No
Stacktrace support       : No
Memory alloc check       : No
libnl version            : 3
Use IPv4 devconf         : No
Use libiptc              : Yes
Use libipset             : Yes
init type                : systemd
Build genhash            : Yes
Build documentation      : No

Tried running the below command, but fails.
make rpm
rpmbuild -ba --build-in-place keepalived.spec
rpmbuild: --build-in-place: unknown option
make: *** [rpm] Error 1

Hence i just ran the rpmbuild command:
rpmbuild  -v -bb --clean keepalived.spec

make[2]: Entering directory `/root/rpmbuild/BUILD/keepalived-1.3.5/keepalived/vrrp'
  CC       vrrp_print.o
  CC       vrrp_daemon.o
  CC       vrrp_data.o
  CC       vrrp_parser.o
  CC       vrrp.o
  CC       vrrp_scheduler.o
  CC       vrrp_notify.o
  CC       vrrp_sync.o
  CC       vrrp_index.o
  CC       vrrp_arp.o
  CC       vrrp_if.o
  CC       vrrp_track.o
  CC       vrrp_ipaddress.o
  CC       vrrp_ndisc.o
  CC       vrrp_if_config.o
  CC       vrrp_vmac.o
  CC       vrrp_ipsecah.o
  CC       vrrp_iproute.o
  CC       vrrp_iprule.o
  CC       vrrp_ip_rule_route_parser.o
  CC       vrrp_iptables.o
  CC       vrrp_iptables_calls.o
  CC       vrrp_ipset.o
vrrp_ipset.c:53:50: error: unknown type name 'ipset_outfn'
 struct ipset_session* (*ipset_session_init_addr)(ipset_outfn outfn);
                                                  ^
vrrp_ipset.c: In function 'create_sets':
vrrp_ipset.c:64:30: error: 'ipset_session_init_addr' undeclared (first use in this function)
 #define ipset_session_init (*ipset_session_init_addr)
                              ^
vrrp_ipset.c:153:12: note: in expansion of macro 'ipset_session_init'
  session = ipset_session_init(printf);
            ^
vrrp_ipset.c:64:30: note: each undeclared identifier is reported only once for each function it appears in
 #define ipset_session_init (*ipset_session_init_addr)
                              ^
vrrp_ipset.c:153:12: note: in expansion of macro 'ipset_session_init'
  session = ipset_session_init(printf);
            ^
vrrp_ipset.c: In function 'ipset_init':
vrrp_ipset.c:217:8: error: 'ipset_session_init_addr' undeclared (first use in this function)
  if (!(ipset_session_init_addr = dlsym(libipset_handle, "ipset_session_init"))) {log_message(LOG_INFO, "Failed to dynamic link ipset_session_init"); err = true;}
        ^
vrrp_ipset.c: In function 'remove_ipsets':
vrrp_ipset.c:64:30: error: 'ipset_session_init_addr' undeclared (first use in this function)
 #define ipset_session_init (*ipset_session_init_addr)
                              ^
vrrp_ipset.c:244:12: note: in expansion of macro 'ipset_session_init'
  session = ipset_session_init(printf);
            ^
vrrp_ipset.c: In function 'ipset_session_start':
vrrp_ipset.c:64:30: error: 'ipset_session_init_addr' undeclared (first use in this function)
 #define ipset_session_init (*ipset_session_init_addr)
                              ^
vrrp_ipset.c:270:9: note: in expansion of macro 'ipset_session_init'
  return ipset_session_init(NULL);
         ^
vrrp_ipset.c:271:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
make[2]: *** [vrrp_ipset.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: Leaving directory `/root/rpmbuild/BUILD/keepalived-1.3.5/keepalived/vrrp'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/root/rpmbuild/BUILD/keepalived-1.3.5/keepalived'
make: *** [all-recursive] Error 1
error: Bad exit status from /var/tmp/rpm-tmp.ZYw66x (%build)
RPM build errors:
    Bad exit status from /var/tmp/rpm-tmp.ZYw66x (%build)

Any help will be much appreciated. I guess i am just missing may kernel params.
I have also tried to uninstall the snmp related packages from the build server, hoping it may be clashing in the linking process.

thanks in advance.
springleo.
[rpmbuild.log](https://github.com/acassen/keepalived/files/4336951/rpmbuild.log)

**Comment by pqarmitage** [comment]
As it says above:
```
rpmbuild: --build-in-place: unknown option
make: *** [rpm] Error 1
```

**Comment by springleo** [comment]
> 
> 
> As it says above:
> 
> ```
> rpmbuild: --build-in-place: unknown option
> make: *** [rpm] Error 1
> ```

"--build-in-place" is just an argument, that got rid of. however, it still fails to make.
`  CC       vrrp_ip_rule_route_parser.o
  CC       vrrp_iptables.o
  CC       vrrp_iptables_calls.o
  CC       vrrp_ipset.o
vrrp_ipset.c:53:50: error: unknown type name 'ipset_outfn'
 struct ipset_session* (*ipset_session_init_addr)(ipset_outfn outfn);
                                                  ^
vrrp_ipset.c: In function 'create_sets':
vrrp_ipset.c:64:30: error: 'ipset_session_init_addr' undeclared (first use in this function)
 #define ipset_session_init (*ipset_session_init_addr)
                              ^
vrrp_ipset.c:153:12: note: in expansion of macro 'ipset_session_init'
  session = ipset_session_init(printf);
            ^
vrrp_ipset.c:64:30: note: each undeclared identifier is reported only once for each function it appears in
 #define ipset_session_init (*ipset_session_init_addr)
                              ^
vrrp_ipset.c:153:12: note: in expansion of macro 'ipset_session_init'
  session = ipset_session_init(printf);
            ^
vrrp_ipset.c: In function 'ipset_init':
vrrp_ipset.c:217:8: error: 'ipset_session_init_addr' undeclared (first use in this function)
  if (!(ipset_session_init_addr = dlsym(libipset_handle, "ipset_session_init"))) {log_message(LOG_INFO, "Failed to dynamic link ipset_session_init"); err = true;}
        ^
vrrp_ipset.c: In function 'remove_ipsets':
vrrp_ipset.c:64:30: error: 'ipset_session_init_addr' undeclared (first use in this function)
 #define ipset_session_init (*ipset_session_init_addr)
                              ^
vrrp_ipset.c:244:12: note: in expansion of macro 'ipset_session_init'
  session = ipset_session_init(printf);
            ^
vrrp_ipset.c: In function 'ipset_session_start':
vrrp_ipset.c:64:30: error: 'ipset_session_init_addr' undeclared (first use in this function)
 #define ipset_session_init (*ipset_session_init_addr)
                              ^
vrrp_ipset.c:270:9: note: in expansion of macro 'ipset_session_init'
  return ipset_session_init(NULL);
         ^
vrrp_ipset.c:271:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
make[3]: *** [vrrp_ipset.o] Error 1
make[3]: *** Waiting for unfinished jobs....
make[3]: Leaving directory `/root/rpmbuild/BUILD/keepalived-1.3.5/keepalived/vrrp'
make[2]: *** [all-recursive] Error 1
make[2]: Leaving directory `/root/rpmbuild/BUILD/keepalived-1.3.5/keepalived'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/root/rpmbuild/BUILD/keepalived-1.3.5'
error: Bad exit status from /var/tmp/rpm-tmp.KXBiyP (%build)
`

**Comment by springleo** [comment]
downgraded the ipset-devel from 7.1.1 to 6.38-2, after which it worked. thanks.

**Comment by pqarmitage** [comment]
It turns out that the tarball used for building keepalived-1.3.5-16_el7 is 4 commits after the v1.3.5 tag in the keepalived repo. Also, if you are attempting to build 1.3.5-16_el7 there are 21 further patches that you would need to apply. However, building the rpm from the contents of the source rpm fails on Centos 7.7 with the same errors that you are getting. It appears that RedHat backported some new ipset kernel functionality into their 3.10+ kernel, and keepalived v1.3.5 does not build against that new functionality.

keepalived v2.0.20 (the current version) does successfully build on Centos 7.7 so you are probably best off using the most recent version of keepalived.

**Comment by springleo** [comment]
Hi, I will need to build the version 1.3.5-16 version only, as that's what we need as part of a compliance. Not my choice. To get this version built as expected, do I need to also downgrade the kernel package? And other libraries too, that I need to downgrade? Please note, I am already able to build by just downgrading the ipset package, but I just want to be doubly sure. Thanks in advance.

**Comment by pqarmitage** [comment]
@springleo Many apologies for the delay in replying.

A bit of background may help here.

RHEL7 was originally released with keepalived v1.3.5, ipset v6 and kernel 3.10. Some time later, RedHat backported ipset V7 kernel functionality (introduced in Linux 5.0) into their RHEL kernel 3.10.0-1018 and consequently upgraded ipset/libipset to v7. Unfortunately RedHat did not upgrade keepalived at the time (RedHat's policy is not to upgrade versions of packages within a release, although ipset appears to be an exception here), and consequently the versions of keepalived and ipset/libipset became incompatible.

I think you have three options if you have to continue using 1.3.5-16:
1. Rollback ipset/libipset as you have done.
2. Build keepalived without ipset support (do you know if you use it?)
3. Backport the keepalived commits b7a98f926 and cf605b912 for working with libipset v7.

Issues:
1. This perhaps has the greatest impact on the system, and may affect other applications that expect ipset/libipset v7.
2. This option is fine if you do not use ipset functionality in keepalived and has the benefit of leaving the rest of the system in a 'standard' state'.
3. If backporting the commits mentioned above works for you this is probably the safest option since you will be running a standard version of RHEL 7, with the changes to keepalived that should be incorporated when moving to libipset v7. These changes have been in use in keepalived since November 2018.

I hope the above helps.
